### DISCUSSION TITLE
How to register Java 17 type handlers with the Helidon SE MicroStream adapter?

### DISCUSSION THREAD

**Discussion by hrstoyanov** [microstream-one/microstream#629] [General]
Hi all, @hg-ms

After looking the example for Helidon SE, I wrote the below configuration code, but it does not look right to me - how is the foundation object affecting the creation of the embedded storage manager builder at all? Am I missing something?

```Java
// Helidon configuration
var config = Config.builder()
                .sources(
                     file(configFile).optional(), 
                     classpath("application.yaml"))
                .build();

// Configure the storage manager
var embeddedStorageManager = EmbeddedStorageManagerBuilder
                .builder()
                .config(config.get("microstream"))
                .build();

// How is this related at all?
final var foundation = EmbeddedStorage.Foundation();
foundation.onConnectionFoundation(BinaryHandlersJDK17::registerJDK17TypeHandlers);
```

Thank you!

**Comment by hg-ms** [comment]
To use those handlers the microstream setup must be done without the help of the EmbeddedStorageManagerBuilder.
 
Please be aware that the helidon integration still uses microstream version 05.00.02-MS-GA.
The BinaryHandlersJDK17 are available since version 06.00.00-MS-GA.
To use the BinaryHandlersJDK17 you should override the version for all microstream dependencies
E.g.: add the property `<version.lib.microstream>08.01.01-MS-GA</version.lib.microstream>` to your pom.xml to use the latest microstream version. 
 
In the first step you may extract the microstream configuration properties from the helidon configuration:
```
Map<String, String> micostreamProperties =  config.get("microstream").detach().asMap().get();
```

Now you can create the EmbeddedStorageFoundation using the properties:
```
EmbeddedStorageConfigurationBuilder configurationBuilder = EmbeddedStorageConfigurationBuilder.New();
micostreamProperties.forEach(configurationBuilder::set);

EmbeddedStorageFoundation<?> foundation = configurationBuilder.createEmbeddedStorageFoundation();
```
Next setup the BinaryHandlersJDK17:
```
foundation.onConnectionFoundation(BinaryHandlersJDK17::registerJDK17TypeHandlers);
```

With the foundation set up, create the EmbeddedStorageManager:
```
EmbeddedStorageManager embeddedStorageManager = foundation.createEmbeddedStorageManager();
```

**Reply by hrstoyanov** [reply]
Thanks @hg-ms, will try it, **but would it better if the Helidon integration is kept up-to-date**? **I can contribute to that?** I also noticed that the helidon integration brings in all the MS jars, whether they are needed or not?

Fortunately, Gradle allows for dependencies force-overrides in situation exactly like this, so if anyone else using Gradle needs this:
```
    //MicroStream dependencies - the property ${helidonVersion} forces a particular version of MS
    implementation "io.helidon.integrations.microstream:helidon-integrations-microstream:${helidonVersion}"
    implementation ("one.microstream:microstream-storage-embedded") version {  strictly "${microstreamVersion}" }
    implementation ("one.microstream:microstream-storage-embedded-configuration") version {  strictly "${microstreamVersion}" }
    implementation ("one.microstream:microstream-configuration") version {  strictly "${microstreamVersion}" }
    implementation "one.microstream:microstream-persistence-binary-jdk17:${microstreamVersion}"
```
and then just verify you got it right with something like this:
> gradle -p webapp dependencies --configuration runtimeClasspath

**Reply by hg-ms** [reply]
Hello
You are right, updating the Helidon integration would be the best solution.

You’re welcome to update the Microstream version used by Helidon if you like.
To do so you need to commit to the Helidon project, as far as I know this requires signing a contributor agreement.
 
When I tried to update the Microstream version the last time (please see https://github.com/helidon-io/helidon/pull/5344) there was a problem with the java module system in Microstream. This should be fixed in the current version (08.x) but must be verified.
This [branch] (https://github.com/microstream-one/helidon/tree/helidon-2.x) is still in our heldion fork.

If like to do that update just let me know. If not, I need to see when we can do it…

best regards
